wire-ios-request-strategy/Tests/Helpers/TestSetup.swift

```diff
@@ -0,0 +1,47 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireTesting
+import WireTransport
+
+final class TestSetup: NSObject, XCTestObservation {
```

Comment:
Would you be able to give a short explanation to how this works? It's the first time I've seen it.

Reply:
Sure. `TestSetup` is configured to be the test targets [NSPrincipleClass](https://developer.apple.com/documentation/xctest/xctestobservationcenter) via the test targets info.plist. This means it is instantiated before any when the test bundle is loaded:

> If an [NSPrincipalClass](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CocoaKeys.html#//apple_ref/doc/plist/info/NSPrincipalClass) key is declared in the test bundle's Info.plist file, XCTest automatically creates a single instance of that class when the test bundle is loaded. You can use this instance as a place to register observers or do other pretesting global setup before testing for that bundle begins.

`TestSetup` conforms to `XCTestObservation` as is added as an observer to the tests so that we can perform actions on certain events during tests. In this case we observe `testBundleWillStart(_:` which is called once before all tests are run in that test target. Here we configure `BackendInfo` to have some default values and, using a `UserDefault` subclass `TestUserDefaults`, ensure that tests fail any time we attempt to mutate a `BackendInfo` value without having enabled mocking on `BackendInfo` via `BackendInfo.enableMocking`. Take a look at `BackendInfo` to see how this works.

In practice this failing on mutating has been really useful for finding which tests are mutating the `BackendInfo` and therefore need some special attention.

Reply:
@johnxnguyen I'm resolving this conversation to merge. Feel free to re-open it if you would like further details.